hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java

```diff
@@ -0,0 +1,245 @@
+/*
+ *  Copyright (c) 2016 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.uber.hoodie.common.util.collection;
+
+import com.uber.hoodie.common.util.SpillableMapUtils;
+import com.uber.hoodie.exception.HoodieIOException;
+import com.uber.hoodie.exception.HoodieNotSupportedException;
+import org.apache.avro.Schema;
+import org.apache.spark.util.SizeEstimator;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * An external map that spills content to disk when there is insufficient space for it
+ * to grow.
+ *
+ * This map holds 2 types of data structures :
+ *
+ *   (1) Key-Value pairs in a in-memory map
+ *   (2) Key-ValueMetadata pairs in an in-memory map which keeps a marker to the values spilled to disk
+ *
+ * NOTE : Values are only appended to disk. If a remove() is called, the entry is marked removed from the in-memory
+ * key-valueMetadata map but it's values will be lying around in the temp file on disk until the file is cleaned.
+ *
+ * The setting of the spill threshold faces the following trade-off: If the spill threshold is
+ * too high, the in-memory map may occupy more memory than is available, resulting in OOM.
+ * However, if the spill threshold is too low, we spill frequently and incur unnecessary disk
+ * writes.
+ * @param <T>
+ * @param <R>
+ */
+public class ExternalSpillableMap<T,R> implements Map<T,R> {
+
+  // maximum space allowed in-memory for this map
+  final private long maxInMemorySizeInBytes;
+  // current space occupied by this map in-memory
+  private Long currentInMemoryMapSize;
+  // Map to store key-values in memory until it hits maxInMemorySizeInBytes
+  final private Map<T,R> inMemoryMap;
+  // Map to store key-valuemetadata important to find the values spilled to disk
+  final private DiskBasedMap<T,R> diskBasedMap;
+  // Schema used to de-serialize and readFromDisk the records written to disk
+  final private Schema schema;
+  // An estimate of the size of each payload written to this map
+  private volatile long estimatedPayloadSize = 0;
+  // TODO (na) : a dynamic sizing factor to ensure we have space for other objects in memory
```

Comment:
@vinothchandar here

Reply:
Can you expand on what you mean by other objects..

Reply:
Well, the other objects is just he DiskBased metadata Map<String, Metadata>. I did some calculations and say for a parquet of 1GB there are about 400K entries spilled to disk (worst case), then 400K* (averageSizeofKey=36bytes + metadataSize = 4 + 8 + 8 + sizeoffilepath=max(100bytes)) = 60 MB. So it's not huge, we could technically get rid of sizingFactor but kept it to have a way to be conservative.

Reply:
okay makes sense